---
title: "LCA_Star"
output:
   html_document: 
      toc: true
      theme: readable
      highlight: default
---

**Goal**: To demonstrate the effectiveness of the current implementation of LCA* for taxonomic assignment with respect to existing methods of $LCA^2$, Contig LCA, and Simple Voting procedures with respect to classification error and sum of squares of NCBI Tree Distance from prediction and actual taxonomy.

## Simulation using MetaSim and NCBI Genomes

* Downloaded `all.fna.tar.gz` and `summary.txt` from the NCBI's ftp server:ftp://ftp.ncbi.nlm.nih.gov/genomes/ Mar 12, 2014
* `summary.txt` contains some metadata for a number of these genomes:

```
Accession  GenbankAcc	Length	Taxid	ProjectID	TaxName	Replicon	Create Date	Update_Date
NC_000907.1	L42023.1	1830138	71421	57771	Haemophilus influenzae Rd KW20	chromosome 	Oct 19 2001	Sep 11 2013  4:30:20:076PM
```

* this kind of information would be useful for an analysis so I cross referenced all the Accession IDs with the actual `.fna` files that I had in `all.fna.tar.gz` with the following shell command:

```
cat summary.txt  | awk '{print $1}' > u
for s in `cat u`; do var1=`find . -name *${s%.[0-9]}*`; echo $s$'\t'$var1 >> ncbiID_to_file; done
```

* it turns out that this is a fairly comprehensive list 25 genomes were dropped because they were not found in the MetaData

```
grep --perl-regexp "\t$" ncbiID_to_file | wc
      25      25     324
grep --perl-regexp "\t$" ncbiID_to_file | wc
NC_003911.11  
AC_000091.1	
NC_009353.1	
NC_009444.1	
NC_010332.1	
NS_000190.1	
NC_011980.1	
NS_000196.1	
NS_000197.2	
NC_012627.1	
NC_012629.1	
NC_012630.1	
NC_012915.1	
NC_013416.1	
NC_013438.1	
NC_013597.1	
NC_013784.1	
NC_013785.1	
NC_013786.1	
NC_013787.1	
NC_013788.1	
NC_014629.1	
NC_015557.1	
NC_015587.1
```

* this leaves us with 2617 genomes with annotation of 2642, storing these in `ncbiID_to_file.txt`

```
grep --perl-regexp ".*fna" ncbiID_to_file2 | wc
    2617    5234  197129
wc ncbiID_to_file2
    2642    5259  197453 ncbiID_to_file2
grep --perl-regexp ".*fna" ncbiID_to_file2 > ncbiID_to_file.txt
```

## Simulating Genomes

Here I'm trying to get a random subset of the NCBI genomes to simulate contigs.

### MetaSim Setup

First I tried with MetaSim.

* all genomes in `all.fna.tar.gz` were loaded into a MetaSim (Version 0.9.1) database
* these genomes were linked via their GI numbers to their `taxid` via the `gi_taxid_nucl.dmp.gz` obtained March 15 2014 from <ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/>
* a taxon profile was create with all genomes in equal number `ncbi_db_taxon_profile.mprf` for all 5104 sequences (genomes and plasmids) for 2713 had taxids
```
grep -o --perl-regexp "taxid.*[0-9]+" ../resources/ncbi_db_taxon_profile.mprf | uniq | wc
    2713    5426   38611
```
*Note I had to reinstall MetaSim as there was an error when the program was writing the sequences sequences to Disk. Changed settings to write out the raw sequence files rather than compressed .gzip*

#### Basic Simulations

* started with a basic simulation using the 'Exact' simulation setting with 5000 reads (Error Model: Normal(10k,5k)) to give a general idea of the performance with respect to a wide range of length distributions.

```
Settings: Exact
[Simulating LCA_Star]  Simulator Settings:
[Simulating LCA_Star]  Preset Name: Exact
[Simulating LCA_Star]  Number Of Reads / Mate Pairs=5000
[Simulating LCA_Star]  Error Model=Exact

[Simulating LCA_Star]  Exact Error Model DNA Clone Parameters=
[Simulating LCA_Star]  Distribution: Normal
[Simulating LCA_Star]  Mean: 10000.0
[Simulating LCA_Star]  2nd parameter: 100.0

[Simulating LCA_Star]  Combine All Files=false
[Simulating LCA_Star]  Uniform Sequence Weights=false
[Simulating LCA_Star]  Number Of Threads=1
[Simulating LCA_Star]  Write FastA=true
[Simulating LCA_Star]  Compress Output Files=false


[Simulating LCA_Star]  +++ File: NCBI_even.mprf +++
[Simulating LCA_Star]  Generating Reads for `NCBI_even.mprf'
```
*MetaSim really doesn't like it when you set a high standard deviation on its Normal error model.* 

* Struggled with MetaSim long enough. Always through memory errors.

### Getting it done in Python

Wrote my own script `subsample_ncbi.py` quickly sample from a collection of fasta files specified in our `ncbiID_to_file.txt` that we created above.

The following example creates sub-sequences of length 10,000, sampling 10 sequences per file on a random subset of 100 (the random number generator is seeded so that results can be reproducible)

```
python subsample_ncbi.py -i ncbiID_to_file.txt -l 10000 -s 10 -n 100 -o lca_star_test1.fna
```

* lca_star_test1.fna is going to be our first test, we ran it through MetaPathways with the standard settings:

```
Run Date : 2014-03-15 
Nucleotide Quality Control parameters
  min length  180
ORF prediction parameters
  min length	60
  algorithm	prodigal
Amino acid quality control and annotation parameters
  min bit score	20
  min seq length	60
  annotation reference dbs	RefSeq_complete_nr_v62_dec2013
  min BSR	0.4
  max evalue	0.000001
Pathway Tools parameters
  taxonomic pruning 	no
rRNA search/match parameters
  min identity	20
  max evalue	0.000001
  rRNA reference dbs	GREENGENES_gg16S-2013-09-12

```

* `create_analysis_table.py` summarizes results into a Long Table format that can be processed by R:

```
python create_analysis_table.py -i lca_star_test/metapathways_out/lca_star_test1/preprocessed/lca_star_test1.mapping.txt --ft_table lca_star_test/metapathways_out/lca_star_test1/results/annotation_table/functional_and_taxonomic_table.txt --sum_table lca_star_test/ncbi_bacteria_genomes/ncbi/summary.txt --ncbi_tree ~/Dropbox/utilities/tree_update/data/pipeline/ncbi_taxonomy_tree.txt -o test.out.txt
```

* example output:

```
shebop:lca_star_test nielsh$ head first_test.txt 
contig	real	taxa	method	dist
lca_star_test1_378	Desulfovibrio desulfuricans subsp. desulfuricans str. ATCC 27774	Desulfomonas Moore et al. 1976	LCA_Star	3
lca_star_test1_378	Desulfovibrio desulfuricans subsp. desulfuricans str. ATCC 27774	all	Majority	11
lca_star_test1_378	Desulfovibrio desulfuricans subsp. desulfuricans str. ATCC 27774	biota	LCA_Squared	10
lca_star_test1_379	Desulfovibrio desulfuricans subsp. desulfuricans str. ATCC 27774	Vibrio cholinicus	LCA_Star	2
lca_star_test1_379	Desulfovibrio desulfuricans subsp. desulfuricans str. ATCC 27774	Vibrio cholinicus	Majority	2
lca_star_test1_379	Desulfovibrio desulfuricans subsp. desulfuricans str. ATCC 27774	Desulfomonas Moore et al. 1976	LCA_Squared	3
lca_star_test1_372	Desulfovibrio desulfuricans subsp. desulfuricans str. ATCC 27774	Desulfomonas Moore et al. 1976	LCA_Star	3
lca_star_test1_372	Desulfovibrio desulfuricans subsp. desulfuricans str. ATCC 27774	all	Majority	11
lca_star_test1_372	Desulfovibrio desulfuricans subsp. desulfuricans str. ATCC 27774	'Desulfovibrionaceae'	LCA_Squared	4
```

* a function `lcastar.get_distance(taxon, real)` was added to the LCAStar.py class to calculate a simple distance between taxa.
* TODO: come up with a distance of 'divergence' where we penalize predicted lineages that are outside of the true lineage

### Preliminary statistics

Preliminary statistics using the exiting NCBI Tree in the current release of MetaPathways:

#### First test

Here compared our first implementation (March 16 2014) of LCA* using the `lca_star_test1.fasta` above (10,000, sampling 10 sequences per file on a random subset of 100):

```{r}
setwd("/Users/nielsh/Dropbox\ (Personal)/projects/LCAStar")
data <- read.table("test.out.txt", sep="\t", header=T)
data2 <- read.table("lca_star_test2_lin_declined.tsv.txt", sep="\t", header=T)
data$dist <- as.numeric(as.character(data$dist))
library(ggplot2)
theme_set(theme_bw())
ggplot(data, aes(x=dist)) + geom_histogram(binwidth=1, aes(fill=method), alpha=0.6) + facet_wrap(~ method)
ggplot(data, aes(x=method, y=dist)) + geom_boxplot(aes(fill=method), alpha=0.6)
```

* The results are encouraging as LCA star seems to have 'significantly' smaller distances compared to the other two methods, meaning it is more specific. Though this first distance metric a small distance might still be divergent which we would like to keep track off.

* Doing some basic delcation by taxa we see that within taxonomic groups LCA* seems to do better on this small subset, however, in some cases, namely Green sulfur bacteria and actinobacteria LCA* was outperformed by the simple Majority rule. 

* Here it would be interesting to compare their divergence, as it is conseavable that a simple majority make quite have significant divergence when wrong and having extreme wins when it is correct.

```{r}
ggplot(data, aes(x=method, y=dist)) + geom_boxplot(aes(fill=real), alpha=0.6)
```

#### Second test

For a second test we sampled sequences of 10,000 bps using 10 subsamples from 2000 randomly selected genomes. Here, did the same procedure except with significantly more samples.

```
python ../../subsample_ncbi.py -i ../ncbiID_to_file.txt -o lca_test2.fasta -l 10000 -s 10 -n 2000
```



```
python create_analysis_table.py -i lca_star_test/metapathways_out/lca_star_test2/preprocessed/lca_star_test2.mapping.txt --ft_table lca_star_test/metapathways_out/lca_star_test2/results/annotation_table/functional_and_taxonomic_table.txt --sum_table lca_star_test/ncbi_bacteria_genomes/ncbi/summary.txt --ncbi_tree ~/Dropbox/utilities/tree_update/data/pipeline/ncbi_taxonomy_tree.txt -o lca_star_test2_lin.txt
```

```{r}
data2$dist <- as.numeric(as.character(data2$dist))
ggplot(data2, aes(x=dist)) + geom_histogram(binwidth=1, aes(fill=method), alpha=0.6) + facet_grid(real_linage~ method)
ggplot(data2, aes(x=method, y=dist)) + geom_boxplot(aes(fill=method), alpha=0.6)
```

```{r}
ggplot(data2, aes(x=method, y=dist)) + geom_boxplot(aes(fill=real_linage), alpha=0.6)
```

## Weighted Taxonomic Distance (WTD)

* Repeat of the above except with the WTD from the Community Paper

```{r}
data$wtd <- as.numeric(as.character(data$wtd))
ggplot(data, aes(x=wtd)) + geom_histogram(aes(fill=method), alpha=0.6, binwidith=0.1) + facet_wrap(real_linage ~ method)
```

* Again with the latest code

```{r}
data3 <- read.table("lca_star_test2_wtd.txt", sep="\t", header=T, fill=T)
```

